readthedocs/projects/tasks/builds.py

```diff
@@ -556,85 +537,30 @@ def update_build(self, state):
-            project=self.data.project,
-            version=self.data.version,
-            build=self.data.build,
-            environment=self.get_vcs_env_vars(),
-            # Force the ``container_image`` to use one that has the latest
-            # ca-certificate package which is compatible with Lets Encrypt
-            container_image=settings.RTD_DOCKER_BUILD_SETTINGS['os']['ubuntu-20.04']
-        )
-
-        # Environment used for code checkout & initial configuration reading
-        with environment:
-            before_vcs.send(
-                sender=self.data.version,
-                environment=environment,
-            )
-
-            self.setup_vcs(environment)
-            self.data.config = load_yaml_config(version=self.data.version)
-            self.save_build_config()
-            self.update_vcs_submodules(environment)
-
-    def update_vcs_submodules(self, environment):
-        version_repo = self.get_vcs_repo(environment)
-        if version_repo.supports_submodules:
-            version_repo.update_submodules(self.data.config)
-
-    def run_build(self):
-        """Build the docs in an environment."""
-        self.data.build_env = self.data.environment_class(
-            project=self.data.project,
-            version=self.data.version,
-            config=self.data.config,
-            build=self.data.build,
-            environment=self.get_build_env_vars(),
+        # NOTE: what about doing:
+        #
+        # (triggered)
+        # - self.clone()
+        # - self.install()
+        # - self.build()
+        # - self.upload()
+        # (finished)
+
+        self.data.builder = DocumentationBuilder(
+            data=self.data,
         )
 
-        # Environment used for building code, usually with Docker
-        with self.data.build_env:
-            python_env_cls = Virtualenv
-            if any([
-                    self.data.config.conda is not None,
-                    self.data.config.python_interpreter in ('conda', 'mamba'),
-            ]):
-                python_env_cls = Conda
-
-            self.data.python_env = python_env_cls(
-                version=self.data.version,
-                build_env=self.data.build_env,
-                config=self.data.config,
-            )
-
-            # TODO: check if `before_build` and `after_build` are still useful
-            # (maybe in commercial?)
-            #
-            # I didn't find they are used anywhere, we should probably remove them
-            before_build.send(
-                sender=self.data.version,
-                environment=self.data.build_env,
-            )
+        self.update_build(state=BUILD_STATE_CLONING)
+        self.data.builder.vcs()
 
-            self.setup_build()
-            self.build_docs()
+        # Sync tags/branches from VCS repository into Read the Docs' `Version`
+        # objects in the database
+        self.sync_versions(self.data.builder.vcs_repository)
```

Comment:
Seems odd that this isn't using the "builder" -- is this because it's shared w/ the sync_repos task?

Reply:
I thought we could split these responsibilities a bit more actually: get the tags/branches from the "builder" and pass them to the `sync_versions` that will only hit the API.

I noted that already to come back later; not sure if it will be in this PR, tho. See https://github.com/readthedocs/readthedocs.org/pull/9002/files/741098b97dbbe842987db2dc556d4a8bab2bd355#diff-a1de8bb51275b9023d77549c15a85c45c68867cd8257ff48d67b8590d5afc8c6R56-R59